From 057e922cec66e3422f74b4fb95542967eb095726 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Fri, 23 Sep 2022 11:01:28 +0000
Subject: [PATCH 059/595] [Huawei] mm/sharepool: Rebind the numa node when
 fallback to normal pages

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187892

----------------------------------------------

When we allocate memory using SP_HUGEPAGE, we would try normal pages when
there was no enough hugepages. The specified numa node information would
get lost when we fallback to normal pages. The result is that we could
allocate memory from other numa node than what we have specified.

The soultion is to rebind the node before retrying.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 5c436b3d5368..299a3dc12fdb 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2623,8 +2623,15 @@ void *mg_sp_alloc(unsigned long size, unsigned long sp_flags, int spg_id)
 	}
 
 	ret = sp_alloc_mmap_populate(spa, &ac);
-	if (ret && ac.state == ALLOC_RETRY)
+	if (ret && ac.state == ALLOC_RETRY) {
+		/*
+		 * The mempolicy for shared memory is located at backend file, which varies
+		 * between normal pages and huge pages. So we should set the mbind policy again
+		 * when we retry using normal pages.
+		 */
+		ac.have_mbind = false;
 		goto try_again;
+	}
 
 out:
 	sp_alloc_finish(ret, spa, &ac);
-- 
2.17.1

